py3status/modules/velib_metropole.py

```diff
@@ -0,0 +1,292 @@
+
+        for code in self.station_codes:
+            new_station = {}
+
+            # search station
+            station = next(
+                (item for item in data if int(item["station"]["code"]) == int(code)),
+                None,
+            )
+
+            # flat, camel to snake and cast...
+            for k, v in self.py3.flatten_dict(station, delimiter="_").items():
+                new_station[self._camel_to_snake_case(k)] = self._cast_number(v)
+
+            # station_due_date_s: station due date in dateime iso format
+            # station_state_code: station code for thresholds
+            station_due_date = datetime.fromtimestamp(
+                int(new_station["station_due_date"])
+            )
+            new_station.update(
+                {
+                    "station_due_date_s": station_due_date.isoformat(),
+                    "station_state_code": int(
+                        self.station_states[new_station["station_state"]]
+                    ),
+                }
+            )
+
+            stations[index] = new_station
+            index += 1
+
+        return len(stations), stations
+
+    def velib_metropole(self):
+        if self.refresh is True:
+            # get number of stations and station list
+            json = self._get_velib_data()
+            if json:
+                self.number_of_stations, self.stations = self._manipulate(json)
+
+        # reset refresh
+        self.refresh = True
+
+        if not self.stations:
+            full_text = None
+        else:
+            # reset station_index counter
+            if self.station_index == 0:
+                self.station_index = 1
+
+            # thresholds
+            for x in self.thresholds:
+                if x in self.stations[self.station_index]:
+                    self.py3.threshold_get_color(
+                        self.stations[self.station_index][x], x
+                    )
+
+            # forge data output
+            velib_data = {
+                "stations": self.number_of_stations,
+                "format_station": self.py3.safe_format(
+                    self.format_station, self.stations[self.station_index]
+                ),
+                "index": self.station_index,
+            }
+            full_text = self.py3.safe_format(self.format, velib_data)
+
+        return {
+            "cached_until": self.py3.time_in(self.cache_timeout),
+            "full_text": full_text,
+        }
+
+    def on_click(self, event):
+        button = event["button"]
+        if self.stations:
+            self.refresh = False
+            if button == self.button_next:
+                self.station_index += 1
+                self.station_index %= self.number_of_stations + 1
+            elif button == self.button_previous:
+                self.station_index -= 1
+                self.station_index %= self.number_of_stations + 1
+        elif button == self.button_refresh:
+            self.refresh = True
```

Comment:
(Opinion) This `self.refresh` code seems a little icky... and could lead to errors somewhere... See `whatismyip` for the nice example as it would track your `self.cache_timeout` and  allow you to scroll away without refreshes.